File: tv_android/app/src/main/java/com/nascabos/tv/modules/video/VideoListApiService.kt
package com.nascabos.tv.modules.video

import com.nascabos.tv.core.api.ApiController

object VideoListApiService {
    private fun authHeaders(): Map<String, String> {
        val token = ApiController.accessToken.trim()
        if (token.isEmpty()) return emptyMap()
        return mapOf("Authorization" to "Bearer $token")
    }

    suspend fun getIndexCounts(
        sourceList: List<String>? = null,
        timeoutSeconds: Long = 10,
    ): VideoIndexCountResult? {
        val body = mutableMapOf<String, Any?>()
        if (!sourceList.isNullOrEmpty()) body["sourceList"] = sourceList

        val raw =
            ApiController.postJsonMap(
                baseUrl = ApiController.baseUrl,
                path = "/api/video/list/count",
                body = body,
                timeoutSeconds = timeoutSeconds,
                headers = authHeaders(),
            )
        val data = unwrapData(raw) ?: return null
        val movie = parseInt(data["movie"])
        val tv = parseInt(data["tv"])
        val total = parseInt(data["total"]).takeIf { it > 0 } ?: (movie + tv)
        return VideoIndexCountResult(movie = movie, tv = tv, total = total)
    }

    suspend fun listPaged(
        page: Int,
        pageSize: Int,
        mediaType: String,
        search: String?,
        sourceList: List<String>?,
        sort: VideoListSortState,
        listType: String? = null,
        albumId: Int? = null,
        collectionId: Int? = null,
        smartAlbumId: Int? = null,
        timeoutSeconds: Long = 15,
    ): VideoListPagedResult {
        val effectivePage = if (page > 0) page else 1
        val effectivePageSize = if (pageSize > 0) pageSize else 30
        val mt = mediaType.trim().lowercase()
        val body =
            mutableMapOf<String, Any?>(
                "page" to effectivePage,
                "page_size" to effectivePageSize,
                "sort_by" to sort.by.apiValue,
                "sort_order" to sort.order.apiValue,
            )
        if (mt == "movie" || mt == "tv" || mt == "season") {
            body["media_type"] = mt
        }
        val q = search?.trim().orEmpty()
        if (q.isNotEmpty()) body["search"] = q
        if (!sourceList.isNullOrEmpty()) body["sourceList"] = sourceList
        val lt = listType?.trim().orEmpty()
        if (lt.isNotEmpty()) body["listType"] = lt
        val aid = albumId ?: 0
        if (aid > 0) body["album_id"] = aid
        val cid = collectionId ?: 0
        if (cid > 0) body["collection_id"] = cid
        val sid = smartAlbumId ?: 0
        if (sid > 0) body["smart_album_id"] = sid

        val raw =
            ApiController.postJsonMap(
                baseUrl = ApiController.baseUrl,
                path = "/api/video/list",
                body = body,
                timeoutSeconds = timeoutSeconds,
                headers = authHeaders(),
            )
        val data = unwrapData(raw) ?: return VideoListPagedResult.empty

        val items = parseItems(data["items"])
        val pagination = parsePagination(data["pagination"])
        val validPaths = parseValidPaths(data["validPaths"] ?: data["valid_paths"])
        return VideoListPagedResult(items = items, pagination = pagination, validPaths = validPaths)
    }

    suspend fun listHistory(
        timeoutSeconds: Long = 20,
    ): List<VideoListItem> {
        val raw =
            ApiController.postJsonMap(
                baseUrl = ApiController.baseUrl,
                path = "/api/video/history/list",
                body = emptyMap<String, Any?>(),
                timeoutSeconds = timeoutSeconds,
                headers = authHeaders(),
            )
        val data = unwrapData(raw) ?: return emptyList()
        return parseItems(data["items"])
    }

    private fun unwrapData(raw: Map<String, Any?>): Map<String, Any?>? {
        val success = raw["success"] as? Boolean
        if (success != null) {
            if (success != true) return null
            val data = raw["data"] as? Map<*, *>
            @Suppress("UNCHECKED_CAST")
            return data?.entries?.associate { it.key?.toString().orEmpty() to it.value } as? Map<String, Any?>
        }
        return raw
    }

    private fun parseInt(v: Any?): Int {
        return when (v) {
            null -> 0
            is Number -> v.toInt()
            else -> v.toString().trim().removeSuffix(".0").toIntOrNull() ?: 0
        }
    }

    private fun parseDouble(v: Any?): Double {
        return when (v) {
            null -> 0.0
            is Number -> v.toDouble()
            else -> v.toString().trim().toDoubleOrNull() ?: 0.0
        }
    }

    private fun parseBool(v: Any?): Boolean {
        return when (v) {
            true -> true
            false -> false
            is Number -> v.toInt() != 0
            else -> v?.toString()?.trim() == "1"
        }
    }

    private fun parseItems(v: Any?): List<VideoListItem> {
        val list = v as? List<*> ?: return emptyList()
        return list.mapNotNull { raw ->
            val map = raw as? Map<*, *> ?: return@mapNotNull null
            val id = parseInt(map["id"])
            if (id <= 0) return@mapNotNull null
            val mediaType = map["media_type"]?.toString().orEmpty()
            val path = map["path"]?.toString().orEmpty()
            val filename = map["filename"]?.toString().orEmpty()
            val firstFilePath = (map["first_file_path"] ?: map["firstFilePath"])?.toString().orEmpty()
            val nfoName = map["nfo_name"]?.toString().orEmpty()
            val nfoYear = parseInt(map["nfo_year"])
            val nfoScore = parseDouble(map["nfo_score"])
            val nfoRegions = map["nfo_regions"]?.toString().orEmpty()
            val nfoGenres = map["nfo_genres"]?.toString().orEmpty()
            val posterPath = map["poster_path"]?.toString().orEmpty()
            val fanartPath = map["fanart_path"]?.toString().orEmpty()
            val logoPath = map["logo_path"]?.toString().orEmpty()
            val progress = parseDouble(map["progress"])
            val isFavorite = parseBool(map["is_favorite"])
            val viewTime = map["view_time"]?.toString()
            val createTime = map["create_time"]?.toString()
            val fullPath = map["full_path"]?.toString().orEmpty()
            VideoListItem(
                id = id,
                mediaType = mediaType,
                path = path,
                filename = filename,
                firstFilePath = firstFilePath,
                nfoName = nfoName,
                nfoYear = nfoYear,
                nfoScore = nfoScore,
                nfoRegions = nfoRegions,
                nfoGenres = nfoGenres,
                posterPath = posterPath,
                fanartPath = fanartPath,
                logoPath = logoPath,
                progress = progress,
                isFavorite = isFavorite,
                viewTime = viewTime,
                createTime = createTime,
                fullPath = fullPath,
            )
        }
    }

    private fun parsePagination(v: Any?): VideoListPagination {
        val map = v as? Map<*, *> ?: return VideoListPagination.empty
        val total = parseInt(map["total"])
        val page = parseInt(map["page"]).takeIf { it > 0 } ?: 1
        val limit = parseInt(map["limit"]).takeIf { it > 0 } ?: 30
        val totalPages = parseInt(map["totalPages"])
        val hasNextPage = (map["hasNextPage"] as? Boolean) ?: parseBool(map["hasNextPage"])
        val hasPrevPage = (map["hasPrevPage"] as? Boolean) ?: parseBool(map["hasPrevPage"])
        return VideoListPagination(
            total = total,
            page = page,
            limit = limit,
            totalPages = totalPages,
            hasNextPage = hasNextPage,
            hasPrevPage = hasPrevPage,
        )
    }

    private fun parseValidPaths(v: Any?): List<VideoListPathItem> {
        val list = v as? List<*> ?: return emptyList()
        return list.mapNotNull { raw ->
            when (raw) {
                is String -> {
                    val p = raw.trim()
                    if (p.isEmpty()) null else VideoListPathItem(path = p, valid = true)
                }
                is Map<*, *> -> {
                    val path = raw["path"]?.toString()?.trim().orEmpty()
                    if (path.isEmpty()) return@mapNotNull null
                    val valid = raw["valid"] as? Boolean ?: parseBool(raw["valid"])
                    VideoListPathItem(path = path, valid = valid)
                }
                else -> null
            }
        }
    }
}

File: flutter_client/lib/modules/video/list/service/video_list_api_service.dart
import 'package:get/get.dart';
import '../../../../core/api/base_api_service.dart';
import '../../base/beans/video_item_bean.dart';

class VideoListApiService extends BaseApiService {
  static VideoListApiService get instance =>
      Get.isRegistered<VideoListApiService>()
      ? Get.find<VideoListApiService>()
      : VideoListApiService();

  Future<ApiResponse<VideoIndexCountResult>> getIndexCounts({
    List<String>? sourceList,
  }) async {
    final res = await apiPost<Map<String, dynamic>>(
      '/api/video/list/count',
      body: {if (sourceList != null) 'sourceList': sourceList},
      showLoading: false,
    );

    if (!res.success) {
      return ApiResponse.failure(
        res.message ?? 'network_failure',
        code: res.code,
        rawResponse: res.rawResponse,
      );
    }

    final data = res.data ?? <String, dynamic>{};
    final parsed = VideoIndexCountResult.fromJson(data);

    return ApiResponse.success(
      parsed,
      message: res.message,
      code: res.code,
      rawResponse: res.rawResponse,
    );
  }

  Future<VideoListPagedResult> listPaged({
    required int page,
    required int pageSize,
    String? listType,
    String? mediaType,
    int? albumId,
    int? collectionId,
    int? smartAlbumId,
    String? search,
    List<String>? genres,
    List<String>? regions,
    List<String>? actors,
    List<String>? directors,
    List<int>? years,
    List<String>? sourceList,
    String? sortBy,
    String? sortOrder,
    bool showLoading = false,
  }) async {
    final res = await apiPost<Map<String, dynamic>>(
      '/api/video/list',
      body: {
        'page': page,
        'page_size': pageSize,
        if (listType != null && listType.trim().isNotEmpty)
          'listType': listType.trim(),
        if (mediaType != null && mediaType.trim().isNotEmpty)
          'media_type': mediaType.trim(),
        if (albumId != null && albumId > 0) 'album_id': albumId,
        if (collectionId != null && collectionId > 0)
          'collection_id': collectionId,
        if (smartAlbumId != null && smartAlbumId > 0)
          'smart_album_id': smartAlbumId,
        if (search != null) 'search': search,
        if (genres != null) 'genres': genres,
        if (regions != null) 'regions': regions,
        if (actors != null) 'actors': actors,
        if (directors != null) 'directors': directors,
        if (years != null) 'years': years,
        if (sourceList != null) 'sourceList': sourceList,
        if (sortBy != null) 'sort_by': sortBy,
        if (sortOrder != null) 'sort_order': sortOrder,
      },
      showLoading: showLoading,
    );

    if (!res.success) return const VideoListPagedResult.empty();
    final data = res.data ?? <String, dynamic>{};
    return VideoListPagedResult.fromJson(data);
  }
}

class VideoIndexCountResult {
  final int movie;
  final int tv;

  const VideoIndexCountResult({required this.movie, required this.tv});

  int get total => movie + tv;

  factory VideoIndexCountResult.fromJson(Map<String, dynamic> json) {
    final rawMovie = json['movie'];
    final rawTv = json['tv'];
    final movie = rawMovie is num
        ? rawMovie.toInt()
        : int.tryParse(rawMovie?.toString() ?? '') ?? 0;
    final tv = rawTv is num
        ? rawTv.toInt()
        : int.tryParse(rawTv?.toString() ?? '') ?? 0;
    return VideoIndexCountResult(movie: movie, tv: tv);
  }
}

class VideoListPagination {
  final int total;
  final int page;
  final int limit;
  final int totalPages;
  final bool hasNextPage;
  final bool hasPrevPage;

  const VideoListPagination({
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
    required this.hasNextPage,
    required this.hasPrevPage,
  });

  const VideoListPagination.empty()
    : total = 0,
      page = 1,
      limit = 30,
      totalPages = 0,
      hasNextPage = false,
      hasPrevPage = false;

  factory VideoListPagination.fromJson(Map<String, dynamic> json) {
    return VideoListPagination(
      total: (json['total'] as num?)?.toInt() ?? 0,
      page: (json['page'] as num?)?.toInt() ?? 1,
      limit: (json['limit'] as num?)?.toInt() ?? 30,
      totalPages: (json['totalPages'] as num?)?.toInt() ?? 0,
      hasNextPage: (json['hasNextPage'] as bool?) ?? false,
      hasPrevPage: (json['hasPrevPage'] as bool?) ?? false,
    );
  }
}

class VideoListFilterOptions {
  final List<String> genres;
  final List<String> regions;
  final List<int> years;

  const VideoListFilterOptions({
    required this.genres,
    required this.regions,
    required this.years,
  });

  factory VideoListFilterOptions.fromJson(Map<String, dynamic> json) {
    List<String> parseStringList(dynamic v) {
      final list = v is List ? v : const <dynamic>[];
      return list
          .map((e) => e.toString())
          .map((e) => e.trim())
          .where((e) => e.isNotEmpty)
          .toList();
    }

    List<int> parseIntList(dynamic v) {
      final list = v is List ? v : const <dynamic>[];
      return list
          .map((e) => e is num ? e.toInt() : int.tryParse(e.toString()) ?? 0)
          .where((e) => e > 0)
          .toList();
    }

    return VideoListFilterOptions(
      genres: parseStringList(json['genres']),
      regions: parseStringList(json['regions']),
      years: parseIntList(json['years']),
    );
  }
}

class VideoListPathItem {
  final String path;
  final bool valid;

  const VideoListPathItem({required this.path, required this.valid});

  factory VideoListPathItem.fromJson(Map<String, dynamic> json) {
    return VideoListPathItem(
      path: (json['path']?.toString() ?? '').trim(),
      valid: (json['valid'] as bool?) ?? false,
    );
  }
}

class VideoListPagedResult {
  final List<VideoHomeItemBean> items;
  final VideoListPagination pagination;
  final VideoListFilterOptions? filters;
  final List<VideoListPathItem> validPaths;

  const VideoListPagedResult({
    required this.items,
    required this.pagination,
    required this.filters,
    required this.validPaths,
  });

  const VideoListPagedResult.empty()
    : items = const <VideoHomeItemBean>[],
      pagination = const VideoListPagination.empty(),
      filters = null,
      validPaths = const <VideoListPathItem>[];

  factory VideoListPagedResult.fromJson(Map<String, dynamic> json) {
    List<VideoHomeItemBean> parseItems(dynamic v) {
      final list = v is List ? v : const <dynamic>[];
      return list
          .whereType<Map>()
          .map((e) => VideoHomeItemBean.fromJson(e.cast<String, dynamic>()))
          .where((e) => e.id > 0)
          .toList();
    }

    List<VideoListPathItem> parseValidPaths(dynamic v) {
      final list = v is List ? v : const <dynamic>[];
      return list
          .map((e) {
            if (e is String) {
              final p = e.trim();
              return VideoListPathItem(path: p, valid: true);
            }
            if (e is Map) {
              return VideoListPathItem.fromJson(e.cast<String, dynamic>());
            }
            return const VideoListPathItem(path: '', valid: false);
          })
          .where((e) => e.path.isNotEmpty)
          .toList();
    }

    final paginationRaw = json['pagination'];
    final paginationMap = paginationRaw is Map
        ? paginationRaw.cast<String, dynamic>()
        : <String, dynamic>{};
    final filtersRaw = json['filters'] ?? json['filterOptions'];
    final filtersMap = filtersRaw is Map
        ? filtersRaw.cast<String, dynamic>()
        : null;
    final validPathsRaw = json['validPaths'] ?? json['valid_paths'];
    return VideoListPagedResult(
      items: parseItems(json['items']),
      pagination: VideoListPagination.fromJson(paginationMap),
      filters: filtersMap == null
          ? null
          : VideoListFilterOptions.fromJson(filtersMap),
      validPaths: parseValidPaths(validPathsRaw),
    );
  }
}
